### PR TITLE
cachegroupparameters rewrite

### LINKED ISSUE #3809: Rewrite /cachegroupparameters to Go
## I'm submitting an ...
improvement request (usability, performance, tech debt, etc.)

## Traffic Control components affected ...
- Documentation (possibly)
- Traffic Ops

## Current behavior:
`/cachegroupparameters` is handled by Perl

## Expected / new behavior:
`/cachegroupparameters` should be handled by Go.

### Minimum API Version
1.1

### Request Methods
- [x] `GET`
- [x] `POST`

## Minimal reproduction of the problem with instructions:
N/A

### PR DESCRIPTION
**Pull request by mattjackson220**
## What does this PR (Pull Request) do?

Rewrites /cachegroupparameters in go
- [x] This PR fixes #3809 

## Which Traffic Control components are affected by this PR?

- Documentation
- Traffic Ops

## What is the best way to verify this PR?
- Verify that a GET to /cachegroupparameters returns a list of all of the cachegroup / parameter associations
- Verify that a PUT to /cachegroupparameters with only one association like so:
{"cachegroupId":1, "parameterId":2}
adds the association to the table and appears with a GET
- Verify that a PUT to /cachegroupparameters with multiple associations like so:
[{"cachegroupId":1, "parameterId":2}, {"cachegroupId":1, "parameterId":2}, {"cachegroupId":2, "parameterId":2}]
adds all of the associations to the table and they all appear with a GET

## The following criteria are ALL met by this PR

- [x] This PR includes tests
- [x] This PR includes documentation
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 1]
```diff
@@ -180,3 +185,120 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+		return
+	}
+
+	values := []string{}
+	for _, param := range params {
+		values = append(values, "("+strconv.Itoa(*param.CacheGroup)+", "+strconv.Itoa(*param.Parameter)+")")
+	}
+
+	insQuery := strings.Join(values, ", ")
+	_, err = inf.Tx.Tx.Query(insertQuery() + insQuery)
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: you should be able (as long as our sqlx is recent enough) to do a bulk insert with a named query so you dont have to convert ints to strings and then joining them before adding them to the query. https://github.com/jmoiron/sqlx/pull/285 is when it was added

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 2]
```diff
@@ -180,3 +185,120 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+
+func AddCacheGroupParameters(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	data, _ := ioutil.ReadAll(r.Body)
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: should we ignore the error here?

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 3]
```diff
@@ -180,3 +185,120 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+func GetAllCacheGroupParameters(tx *sql.Tx) (tc.CacheGroupParametersList, error) {
+	parameters := map[string]string{
+		"orderby": "cachegroup",
+	}
+	// Query Parameters to Database Query column mappings
+	// see the fields mapped in the SQL query
+	queryParamsToQueryCols := map[string]dbhelpers.WhereColumnInfo{
+		"cachegroup":  dbhelpers.WhereColumnInfo{"cachegroup", api.IsInt},
+		"lastUpdated": dbhelpers.WhereColumnInfo{"last_updated", nil},
+		"parameter":   dbhelpers.WhereColumnInfo{"parameter", api.IsInt},
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: this needs to be added to the doc as support request query parameters

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 4]
```diff
@@ -180,3 +185,120 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+	values := []string{}
+	for _, param := range params {
+		values = append(values, "("+strconv.Itoa(*param.CacheGroup)+", "+strconv.Itoa(*param.Parameter)+")")
+	}
+
+	insQuery := strings.Join(values, ", ")
+	_, err = inf.Tx.Tx.Query(insertQuery() + insQuery)
+
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("writing cachegroup parameter: "+err.Error()))
```
[the remark this answers is not in the document]
Comment (mhoppa) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: Perl also returned 4xx response code if that parameter was already associated with the cache group with this implementation does not do

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 5]
```diff
@@ -180,3 +185,120 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+		return
+	}
+
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: I also dont see any logic matching -> 
https://github.com/apache/trafficcontrol/blob/master/traffic_ops/app/lib/API/CachegroupParameter.pm#L72-L76

File: lib/go-tc/cachegroup_parameter_associations.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	Alerts
+}
+
+type CacheGroupParametersList struct {
+	CacheGroupParameters []CacheGroupParametersNullable `json:"cachegroupParameters"`
+}
+
+// CacheGroupParameter ...
+type CacheGroupParameters struct {
+	CacheGroup  int       `json:"cachegroupId"  db:"cachegroup"`
```
Comment (mhoppa) [inline lib/go-tc/cachegroup_parameter_associations.go]: FWIW I believe `cachegroupId` was handled as `cacheGroupId` in Perl we should match that

Comment (ocket8888) [reply lib/go-tc/cachegroup_parameter_associations.go]: Yeah, there was a typo in the docs example, which I tried to fix in time for someone rewriting this endpoint to see, but https://github.com/apache/trafficcontrol/pull/4329 still has to get merged....

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 7]
```diff
@@ -180,3 +185,120 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+	api.WriteRespAlertObj(w, r, tc.SuccessLevel, "Cachegroup parameter associations were created.", params)
+}
+
+func selectAllQuery() string {
+	return `SELECT * FROM cachegroup_parameter`
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: done

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: done

File: lib/go-tc/cachegroup_parameter_associations.go (hunk 6 above)
Comment (mattjackson220) [reply lib/go-tc/cachegroup_parameter_associations.go]: done

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go (hunk 2 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: no we shouldnt. updated

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go (hunk 1 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: doesnt look like ours is up to date enough to do this unfortunately

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go (hunk 3 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: on second thought i dont think we want to do this. ill take it out

File: docs/source/api/cachegroupparameters.rst [hunk 8]
```diff
@@ -97,6 +105,32 @@ The request data can take the form of either a single object or an array of one
...
+	POST /api/1.1/cachegroupparameters HTTP/1.1
+	Host: trafficops.infra.ciab.test
+	User-Agent: curl/7.47.0
+	Accept: */*
+	Cookie: mojolicious=...
+	Content-Length: 39
+	Content-Type: application/json
+
+	[{
+		"cachegroupId": 8,
```
Comment (mhoppa) [inline docs/source/api/cachegroupparameters.rst]: this and below need to be `cacheGroupId`

File: lib/go-tc/cachegroup_parameter_associations.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+// CacheGroupParameterNullable ...
+type CacheGroupParametersNullable struct {
+	CacheGroup  *int       `json:"cacheGroupId"  db:"cachegroup"`
+	Parameter   *int       `json:"parameterId"  db:"parameter"`
+	LastUpdated *TimeNoMod `json:"lastUpdated,omitempty"  db:"last_updated"`
+}
+
+// CacheGroupParameterResponseNullable ...
+type CacheGroupParametersResponseNullable struct {
+	CacheGroup  *int       `json:"cachegroup"  db:"cachegroup"`
```
Comment (mhoppa) [inline lib/go-tc/cachegroup_parameter_associations.go]: this should be a string representing the cachegroup name instead of id

File: traffic_ops/client/cachegroup_parameters.go [hunk 10]
```diff
@@ -67,6 +67,28 @@ func (to *Session) getCacheGroupParameters(route, queryParams string) ([]tc.Cach
 	return data.Response, reqInf, nil
 }
 
+// GetAllCacheGroupParameters Gets all Cachegroup Parameter associations
+func (to *Session) GetAllCacheGroupParameters() ([]tc.CacheGroupParametersResponseNullable, ReqInf, error) {
+	route := fmt.Sprintf("%s/", API_v1_CacheGroupParameters)
+	return to.getAllCacheGroupParameters(route, "")
+}
+
+func (to *Session) getAllCacheGroupParameters(route, queryParams string) ([]tc.CacheGroupParametersResponseNullable, ReqInf, error) {
```
Comment (mhoppa) [inline traffic_ops/client/cachegroup_parameters.go]: just curious why the need for the two methods?

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 11]
```diff
@@ -180,3 +185,124 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+		api.WriteRespAlertObj(w, r, tc.ErrorLevel, "querying cachegroupparameters with error: "+err.Error(), output)
+		return
+	}
+	api.WriteResp(w, r, output)
+}
+
+// GetAllCacheGroupParameters gets all cachegroup associations from the database and returns as slice.
+func GetAllCacheGroupParameters(tx *sql.Tx) (tc.CacheGroupParametersList, error) {
+	parameters := map[string]string{
+		"orderby": "cachegroup",
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: why the change from https://github.com/apache/trafficcontrol/blob/f1447bde94e84aa93fba9506a387a34e9aea3dd7/traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go#L210 also the parameters passed into the `BuildWhereAndOrderByAndPagination` should be parameters that the APIInfo parsed out ```inf.Params```

File: docs/source/api/cachegroupparameters.rst (hunk 8 above)
Comment (mattjackson220) [reply docs/source/api/cachegroupparameters.rst]: done

File: lib/go-tc/cachegroup_parameter_associations.go (hunk 9 above)
Comment (mattjackson220) [reply lib/go-tc/cachegroup_parameter_associations.go]: done

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go (hunk 11 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: i was thinking to mirror the perl endpoint better but added it back in as well as docs

File: lib/go-tc/cachegroup_parameter_associations.go (hunk 9 above)
Comment (mattjackson220) [reply lib/go-tc/cachegroup_parameter_associations.go]: done

File: traffic_ops/client/cachegroup_parameters.go (hunk 10 above)
Comment (mattjackson220) [reply traffic_ops/client/cachegroup_parameters.go]: oh originally i was going to mirror the get methods above but realized i only needed the one but left it split even though the second one isnt reused. ill combine them again since theres no need for 2

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go (hunk 11 above)
Comment (ocket8888) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: Lemme throw this out there: you should support a proper superset of the query parameters supported by Perl - more is totally acceptable, if you want, as long as they are documented (along with the version in which they were added).

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 12]
```diff
@@ -180,3 +187,137 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+	}
+
+	api.WriteRespAlertObj(w, r, tc.SuccessLevel, "Cachegroup parameter associations were created.", params)
+}
+
+func selectAllQuery() string {
+	return `SELECT cgp.cachegroup, cgp.parameter, cgp.last_updated, cg.name 
+				FROM cachegroup_parameter AS cgp 
+				JOIN cachegroup AS cg ON cg.id = cachegroup`
+	//return `SELECT cachegroup, parameter, last_updated FROM cachegroup_parameter`
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: remove comment

File: traffic_ops/client/cachegroup_parameters.go [hunk 13]
```diff
@@ -67,6 +67,24 @@ func (to *Session) getCacheGroupParameters(route, queryParams string) ([]tc.Cach
 	return data.Response, reqInf, nil
 }
 
+// GetAllCacheGroupParameters Gets all Cachegroup Parameter associations
+func (to *Session) GetAllCacheGroupParameters() ([]tc.CacheGroupParametersResponseNullable, ReqInf, error) {
+	route := fmt.Sprintf("%s/", API_v1_CacheGroupParameters)
+	r := fmt.Sprintf("%s/%s", route, "")
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply traffic_ops/client/cachegroup_parameters.go]: done

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 14]
```diff
@@ -180,3 +187,137 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+	}
+
+	where, orderBy, pagination, queryValues, errs := dbhelpers.BuildWhereAndOrderByAndPagination(parameters, queryParamsToQueryCols)
+	if len(errs) > 0 {
+		return tc.CacheGroupParametersList{}, util.JoinErrs(errs)
+	}
+
+	log.Errorf("MATT JACKSON queryparams = %s", queryValues)
+	log.Errorf("MATT JACKSON parameters = %s", parameters)
+	log.Errorf("MATT JACKSON query = %s", selectAllQuery()+where+orderBy+pagination)
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: well thats embarrassing. done

File: docs/source/api/cachegroupparameters.rst [hunk 15]
```diff
@@ -29,7 +29,15 @@ Extract information about the :ref:`cache-group-parameters` associated with :ter
...
+.. table:: Request Query Parameters
+
+	+-------------+----------+--------------------------------------------------------------------------------------------------------------------------------------+
+	| Name        | Required | Description                                                                                                                          |
+	+=============+==========+======================================================================================================================================+
+	| cachegroup  | no       | Show only the :ref:`cache-group-parameters` with the :term:`Cache Group` identified by this integral, unique identifier              |
+	+-------------+----------+--------------------------------------------------------------------------------------------------------------------------------------+
+	| parameter   | no       | Show only the :ref:`cache-group-parameters` with the :term:`Parameter` identified by this integral, unique identifier                |
+	+-------------+----------+--------------------------------------------------------------------------------------------------------------------------------------+
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply docs/source/api/cachegroupparameters.rst]: done

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go (hunk 12 above)
Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: done

File: lib/go-tc/cachegroup_parameter_associations.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+ * under the License.
+ */
+
+// CacheGroupParametersResponse is a Cache Group Parameter response body.
+type AllCacheGroupParametersResponse struct {
+	Response CacheGroupParametersList `json:"response"`
+	Alerts
+}
+
+type CacheGroupParametersList struct {
```
[the remark this answers is not in the document]
Comment (mattjackson220) [reply lib/go-tc/cachegroup_parameter_associations.go]: done

File: traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go [hunk 17]
```diff
@@ -180,3 +186,145 @@ func (cgparam *TOCacheGroupParameter) Delete() (error, error, int) {
...
+	}
+	if parseErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("parsing cachegroup parameter: "+parseErr.Error()), nil)
+		return
+	}
+
+	for _, p := range params {
+		ppExists, err := dbhelpers.ProfileParameterExistsByParameterID(*p.Parameter, inf.Tx.Tx)
+		if err != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, nil, err)
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: StatusInternalServer error here

Comment (mattjackson220) [reply traffic_ops/traffic_ops_golang/cachegroupparameter/parameters.go]: oh my bad. done
